Practice

masamune17 opened the pull request:
このリポジトリは FjordBootCamp の Web アプリケーション プラクティスの提出物です。

masamune17 committed:
雛形作成

メイン処理を行う"main.rb"と"views"ディレクトリを作成

masamune17 committed:
top画面とnew画面の作成

new画面でpostしたデータを保持し、top画面に反映させる

masamune17 committed:
URLの変更とボタンの名称変更

'/new'→'/new_memo'に変更
top画面のボタンの名称を'追加'に変更
new_memo画面のボタンの名称を'保存'に変更

masamune17 committed:
fileの作成、書き込み

"/new_memo/add"実行時、"db"ディレクトリファイルを新規作成、書き込みを行う
その際ファイル名は乱数で生成される

masamune17 committed:
cssを反映

masamune17 committed:
class Memo を定義

masamune17 committed:
ハッシュ を用いてデータを格納

masamune17 committed:
jsonファイルへの読み書き、メモの編集、削除機能を追加

masamune17 committed:
css編集

masamune17 committed:
xss対策

masamune17 committed:
リファクタリング

masamune17 committed:
READMEの編集

masamune17 committed:
guiファイルの削除

maedana commented on README.md:
````diff
@@ -0,0 +1,27 @@
+次に gem をインストールしてください。
+
+```shell
+bundle install
````
Gemfileないのでbundleなんて出来ないですね。↑に書いてあるようにグローバルにgemインストールするのはアンチパターンなので、Gemfile作ってbundle出来るようにしましょうー。じゃないとgemファイルのバージョン固定が出来ないです。

masamune17 replied:
Gemfile作って"bundle install"できるよう変更しました。
それに伴い、READMEを編集いたしました。

maedana commented on db/db.json:
```diff
@@ -0,0 +1 @@
+[{"memo_name":"&lt;script&gt;alert(&#x27;XSS&#x27;);&lt;&#x2F;script&gt;","memo_id":"ae13b885-c2fd-4a73-8a05-0a5273bdbb9c","memo_text":"&lt;script&gt;alert(&#x27;XSS&#x27;);&lt;&#x2F;script&gt;"},{"memo_name":"&lt;script&gt;alert(&#x27;XSS&#x27;);&lt;&#x2F;script&gt;","memo_id":"8cba38dd-8f7c-4427-99b7-ae45878816e6","memo_text":"&lt;script&gt;alert(&#x27;XSS&#x27;);&lt;&#x2F;script&gt;"}]
```
データはコミットしない方が自然じゃないですかね?

masamune17 replied:
誤ってデータをコミットしていたので削除いたしました。

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+
+  def self.read_json
+    json_dates = []
+    File.open('db/db.json', 'r') do |file|
```
もしファイルパス変わったときに複数ヶ所直すことになっちゃうので、ファイルパスは一箇所の変更で済むようにDRYにしておきたいですね

masamune17 replied:
ファイルパスを変数にし、一箇所の変更で済むようにいたしました。

main.rb　line18

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+  end
+
+  def self.read_json
+    json_dates = []
```
dates?? dataの複数形のつもりだとすると、dataの時点で複数形だったと思います。単数形はdatumだったような

masamune17 replied:
dates と記載していた箇所を memos に変更いたしました。

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+      json_dates = JSON.load(file)
+    end
+    json_dates = [] if json_dates.nil?
+    json_dates
```
これぐらいなら僕なら一行で
```ruby
File.open('db/db.json', 'r') { |file| JSON.load(file) } || []
```
と書くかもです。ご参考までに。

masamune17 replied:
いただいたコメントのように、コードを変更いたしました。

main.rb line27

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+  @memo_dates << Memo.create_memo(h(params[:memo_name]), SecureRandom.uuid, h(params[:memo_text]))
+  Memo.write_json(@memo_dates)
+  redirect '/top'
+end
```
> RESTに倣ったURIになっていること

を満たしてないですねー。REST復習してURL全体的に見直してみましょうー。

masamune17 replied:
RESTの課題を終える前に作成しておりましたので、RESTを意識せずにURLをつけていました。

RESTを意識したURLに変更いたしました。

maedana commented on public/js/main.js:
```diff
@@ -0,0 +1,9 @@
+  }else if(!memo_name.match(/\S/g)){
+    return false;
+  }
+}
```
必須チェックいれてるのいいですね :+1:

masamune17 replied:
アラートが出るように変更いたしました。

main.js line7

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+
+get '/top' do
+  @title = 'top'
+  @memo_dates = Memo.read_json
```
上述の通りdataは単数形と複数形ややこしくなるので、単に`@memos`でいいんじゃないかと思いました。いかがでしょう?

masamune17 replied:
dates と記載していた箇所を memos に変更いたしました。

maedana commented on views/top.erb:
```diff
@@ -0,0 +1,13 @@
+  <ul class="top_lists">
+    <% @memo_dates.each do |memo_date| %>
+      <li class="top_list">
+        <a href=/show_memo/<%=memo_date["memo_id"]%>><%=memo_date["memo_name"]%></a><br>
```
xss対策を保存時にやっているですが、表示時にやるのがセオリーかなと思います。でないとユーザの入力値とは異なる値(エスケープされた状態)を保存することになってしまいますので。

masamune17 replied:
xss対策を表示時に行うように変更いたしました。

maedana commented on main.rb:
```diff
@@ -0,0 +1,104 @@
+
+get '/show_memo/:id' do
+  @title = 'Show memo'
+  @memo_dates = Memo.read_json
```
viewで`@memo_dates`は使わないですよね? 使わないものはローカル変数にしましょう。ここ以外にもあるのでそれぞれ対処しましょう

masamune17 replied:
ローカル変数に変更しても問題のないインスタンス変数をローカル変数に変更いたしました。

masamune17 committed:
'db/db.json'をDRYに変更

'db/db.json'→json_path

masamune17 committed:
変数名memo_datesをmemosに変更

masamune17 committed:
xss対策を保存時ではなく表示時に実施

masamune17 committed:
RESTを意識したURLに変更ennkou

masamune17 committed:
Gemfileを作成、READMEの編集

maedana commented on Gemfile:
```diff
@@ -0,0 +1,9 @@
+
+git_source(:github) {|repo_name| "https://github.com/#{repo_name}" }
+
+# gem "rails"
```
これいらないコメントだと思うので削りましょう

maedana commented on main.rb:
```diff
@@ -47,58 +44,58 @@ def self.edit_memo(memo_name, memo_id, memo_text)
-patch '/edit_memo/edit_run/:id' do
-  @memo_dates = Memo.read_json
-  @edit_before = @memo_dates.find do |memo_date|
+patch '/memos/:id/edit/run' do
```
作成したメモのリソースって`/memos/:id`なわけですよね。で、どんな操作が出来るかっていうと、getが取得、deleteが削除、patchが更新なわけですよね。getとdeleteは正しいと思うんですが、更新はなぜ違うURIにしたんでしょう?

masamune17 replied:
対象となるリソースが誤っておりましたので、修正いたしました。

main.rb line67 '/memos/:id/edit/run' → '/memos/:id'

maedana commented on main.rb:
```diff
@@ -47,58 +44,58 @@ def self.edit_memo(memo_name, memo_id, memo_text)
-  @memo_dates << Memo.create_memo(h(params[:memo_name]), SecureRandom.uuid, h(params[:memo_text]))
-  Memo.write_json(@memo_dates)
-  redirect '/top'
+post '/memos/create_memo' do
```
これもRESTっぽくないかなって思います。

https://railsguides.jp/routing.html#crud、動詞、アクション

が参考になるかなって思います。フォーム画面表示は厳密にはREST関係ないので独自のURIを割り当てるって感じですね。

masamune17 replied:
URIを修正いたしました。

main.rb line67 '/memos/create_memo' → '/memos'

masamune17 committed:
Gemfileの余分なコメントを削除、URIの変行

main.rb line67 '/memos/:id/edit/run' → '/memos/:id/edit'
main.rb line96 '/memos/create_memo' → '/memos'

masamune17 committed:
URIの変更

main.rb line67 '/memos/:id/edit' → '/memos/:id'

masamune17 committed:
URIを修正

main.rb line67
